pkg/mcp/internal/queue.go

```diff
@@ -0,0 +1,210 @@
+//
+// - Enqueuing an item in the queue creates a schedule event. The caller can select
+//   over the Ready() channel to process this event and remove items from the queue.
+//
+// - The maximum queue depth is fixed.
+//
+// - The queue can be safely closed. The caller is responsible for checking the Done()
+//   state of the queue and should stop checking Ready() and invoking Dequeue() when
+//   the channel returned by Done() is closed.
+//
+// This is intended to be used by the MCP source/server packages for managing
+// per-type watch state.
+type UniqueQueue struct {
+	sync.Mutex
+	doneChan       chan struct{}
+	doneChanClosed bool
+	// Enqueue at the tail, Dequeue from the head
+	// head == tail   => empty
+	// head == tail+1 => full
+	head int
+	tail int
+
+	// Maintain an ordered set of enqueued items.
+	queue     []entry
+	queuedSet map[string]*entry
+	maxDepth  int
+
+	scheduleChan chan struct{}
+}
+
+type entry struct {
+	key string
+	val interface{}
+}
+
+// NewUniqueScheduledQueue creates a new unique queue specialized for MCP source/server implementations.
+func NewUniqueScheduledQueue(maxDepth int) *UniqueQueue {
+	return &UniqueQueue{
+		queue:        make([]entry, maxDepth+1),
+		queuedSet:    make(map[string]*entry, maxDepth),
+		maxDepth:     maxDepth,
+		scheduleChan: make(chan struct{}, maxDepth),
+		doneChan:     make(chan struct{}),
+	}
+}
+
+// return the next index accounting for wrap
+func (q *UniqueQueue) inc(idx int) int {
+	return (idx + 1) % (q.maxDepth + 1)
+}
+
+// Empty returns true if the queue is empty
+func (q *UniqueQueue) Empty() bool {
+	q.Lock()
+	defer q.Unlock()
+	return q.head == q.tail
+}
+
+// internal version of full() that can be called
+// with the lock held.
+func (q *UniqueQueue) full() bool {
+	return q.head == q.inc(q.tail)
+}
+
+// Full returns true if the queue is full
+func (q *UniqueQueue) Full() bool {
+	q.Lock()
+	defer q.Unlock()
+	return q.full()
+}
+
+// Enqueue an item in the queue. Items with the same key may be safely enqueued multiple
+// times. Enqueueing an item with a key that has already queued has no affect on the order
+// of existing queued items.
+//
+// Returns true if the item exists in the queue upon return. Otherwise,
+// returns false if the item could not be queued.
+func (q *UniqueQueue) Enqueue(key string, val interface{}) bool {
+	q.Lock()
+	defer q.Unlock()
+
+	// Key is already in the queue. Update the set's entry and
+	// return without modifying its place in the queue.
+	if entry, ok := q.queuedSet[key]; ok {
+		entry.val = val
+		return true
+	}
+
+	if q.full() {
+		return false
+	}
+
+	q.queue[q.tail] = entry{
+		key: key,
+		val: val,
+	}
+	q.queuedSet[key] = &q.queue[q.tail]
+	q.tail = q.inc(q.tail)
+
+	q.schedule()
+	return true
+}
+
+// Dequeue removes an item from the queue. This should only be called once
+// for each time Ready() indicates a new event is ready to be processed.
+func (q *UniqueQueue) Dequeue() (string, interface{}, bool) {
```

Comment:
The boolean return seems a little redundant, the caller should be able to figure that out using the key and the value.

Reply:
Perhaps. The dequeued value may be nil with a valid key. Having the explicit boolean makes the contract explicit.